onnxruntime/core/optimizer/compute_optimizer/upstream_reshape_actors.h

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// The optimization here ideally applies to both training and inference,
+// while so far we mainly validate training during cooking the optimization.
+#ifdef ENABLE_TRAINING_CORE
+#pragma once
+
+#include "core/optimizer/compute_optimizer/shared_utils.h"
+
+namespace onnxruntime::optimizer::compute_optimizer {
+
+/**
+ * @brief Struct to hold the information of the reshape operations.
+ *
+ * Initially, an instance of this class for the entry node is created, as the Reshape op propagates
+ * to the entry node's inputs, more instances of this class are created. The propagation stops when
+ * all inputs are not supported to be reshaped.
+ */
+struct ReshapeInfo : public UpstreamOperatorInfoBase {
+  static constexpr int kDataInputIndex = 0;
+  static constexpr int kReshapeOutputIndex = 0;
+
+  ReshapeInfo(const Graph& graph, Node* reshape_node,
+              bool is_entry_node_ptr = false)
+      : UpstreamOperatorInfoBase(reshape_node, is_entry_node_ptr) {
+    const NodeArg* input = reshape_node->InputDefs()[kDataInputIndex];
+    ORT_ENFORCE(input->Shape()->dim_size() == 3, "Only support data of 3D");
+    const NodeArg* output = node_ptr->OutputDefs()[kReshapeOutputIndex];
+    last_dim = output->Shape()->dim(0);
+
+    if (is_entry_node_ptr) {
+      entry_reshape_arg_name = node_ptr->OutputDefs()[kReshapeOutputIndex]->Name();
+    }
+
+    const Node* producer = graph.GetProducerNode(input->Name());
+    if (producer) {
+      // Allow the data input to be grpah input or initializer, but this won't be passed through further.
+      data_producer_output_index = optimizer_utils::IndexOfNodeOutput(*producer, *input);
+    }
+  }
+
+  int GetDataInputIndex() const {
+    return kDataInputIndex;
+  }
+
+  int GetOutputIndex() const {
+    return kReshapeOutputIndex;
+  }
+
+  std::string entry_reshape_arg_name;
+
+  // The dimension of the output tensor after merging the two leading dims.
+  ONNX_NAMESPACE::TensorShapeProto_Dimension last_dim;
+
+  int data_producer_output_index;
```

Comment:
data_producer_output_index seems no use in anywhere ?

Reply:
Good catch! We should use it in PreCheck.